examples/nlp/text_extraction_with_bert.py

```diff
@@ -0,0 +1,323 @@
+        dataset_dict["input_ids"],
+        dataset_dict["token_type_ids"],
+        dataset_dict["attention_mask"],
+    ]
+    y = [dataset_dict["start_token_idx"], dataset_dict["end_token_idx"]]
+    return x, y
+
+
+train_squad_examples = create_squad_examples(raw_train_data)
+x_train, y_train = create_inputs_targets(train_squad_examples)
+print(f"{len(train_squad_examples)} training points created.")
+
+eval_squad_examples = create_squad_examples(raw_eval_data)
+x_eval, y_eval = create_inputs_targets(eval_squad_examples)
+print(f"{len(eval_squad_examples)} evaluation points created.")
+
+"""
+Create the Question Answering Model using BERT and Functional API
+"""
+
+
+def create_model():
+    ## BERT encoder
+    encoder = TFBertModel.from_pretrained("bert-base-uncased")
+
+    ## QA Model
+    input_ids = layers.Input(shape=(max_len,), dtype=tf.int32)
+    token_type_ids = layers.Input(shape=(max_len,), dtype=tf.int32)
+    attention_mask = layers.Input(shape=(max_len,), dtype=tf.int32)
+    embedding = encoder(
+        input_ids, token_type_ids=token_type_ids, attention_mask=attention_mask
+    )[0]
+
+    start_logits = layers.Dense(1, name="start_logit", use_bias=False)(embedding)
+    start_logits = layers.Flatten()(start_logits)
+
+    end_logits = layers.Dense(1, name="end_logit", use_bias=False)(embedding)
+    end_logits = layers.Flatten()(end_logits)
+
+    start_probs = layers.Activation(keras.activations.softmax)(start_logits)
+    end_probs = layers.Activation(keras.activations.softmax)(end_logits)
+
+    model = keras.Model(
+        inputs=[input_ids, token_type_ids, attention_mask],
+        outputs=[start_probs, end_probs],
+    )
+    loss = keras.losses.SparseCategoricalCrossentropy(from_logits=False)
+    optimizer = keras.optimizers.Adam(lr=5e-5)
+    model.compile(optimizer=optimizer, loss=[loss, loss])
+    return model
+
+
+use_tpu = True
+if use_tpu:
+    # create distribution strategy
+    tpu = tf.distribute.cluster_resolver.TPUClusterResolver()
+    tf.config.experimental_connect_to_cluster(tpu)
+    tf.tpu.experimental.initialize_tpu_system(tpu)
+    strategy = tf.distribute.experimental.TPUStrategy(tpu)
+
+    # create model
+    with strategy.scope():
+        model = create_model()
+else:
+    model = create_model()
+
+model.summary()
+
+"""
+## Create evaluation Callback
+
+This callback will compute the exact match score using the validation data
+after every epoch.
+"""
+
+
+def normalize_text(text):
+    text = text.lower()
+
+    # remove punctuations
+    exclude = set(string.punctuation)
+    text = "".join(ch for ch in text if ch not in exclude)
+
+    # remove articles
+    regex = re.compile(r"\b(a|an|the)\b", re.UNICODE)
+    text = re.sub(regex, " ", text)
+
+    # remove extra white space
+    text = " ".join(text.split())
+    return text
+
+
+class ExactMatch(keras.callbacks.Callback):
+    def on_epoch_end(self, epoch, logs=None):
+        pred_start, pred_end = self.model.predict(x_eval)
```

Comment:
Rather than using `x_eval` from the outer scope, pass it to `__init__` and set it as callback attribute

In general: data should be passed as argument, functions / classes can be fetched from the outer scope.